drivers/gpio/gpio_psoc6.c

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2018 Justin Watson
+ * Copyright (c) 2020 ATL Electronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT cypress_psoc6_gpio
+
+#include <errno.h>
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <drivers/gpio.h>
+
+#include "gpio_utils.h"
+#include "cy_gpio.h"
+#include "cy_sysint.h"
+
+#define LOG_LEVEL CONFIG_GPIO_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(gpio_psoc6);
+
+typedef void (*config_func_t)(const struct device *dev);
+
+struct gpio_psoc6_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	GPIO_PRT_Type *regs;
+	config_func_t config_func;
+};
+
+struct gpio_psoc6_runtime {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	sys_slist_t cb;
+};
+
+#define DEV_CFG(dev) \
+	((const struct gpio_psoc6_config * const)(dev)->config)
+#define DEV_DATA(dev) \
+	((struct gpio_psoc6_runtime * const)(dev)->data)
+
+static int gpio_psoc6_config(const struct device *dev,
+			     gpio_pin_t pin,
+			     gpio_flags_t flags)
+{
+	const struct gpio_psoc6_config * const cfg = DEV_CFG(dev);
+	GPIO_PRT_Type * const port = cfg->regs;
+	uint32_t drv_mode = 0;
+	uint32_t pin_val = 0;
+
+	if (flags & GPIO_SINGLE_ENDED) {
+		drv_mode = (flags & GPIO_OPEN_DRAIN) ?
+			   CY_GPIO_DM_OD_DRIVESLOW_IN_OFF :
+			   CY_GPIO_DM_OD_DRIVESHIGH_IN_OFF;
+	} else if ((flags & GPIO_PULL_UP) && (flags & GPIO_PULL_DOWN)) {
+		drv_mode |= CY_GPIO_DM_PULLUP_DOWN_IN_OFF;
+	} else if (flags & GPIO_PULL_UP) {
+		drv_mode |= CY_GPIO_DM_PULLUP_IN_OFF;
+		pin_val = 1;
+	} else if (flags & GPIO_PULL_DOWN) {
+		drv_mode |= CY_GPIO_DM_PULLDOWN_IN_OFF;
+		pin_val = 0;
+	}
+
+	if (flags & GPIO_INPUT) {
+		/**
+		 * When a Input is defined, the pin initial value is
+		 * fundamental to system proper work!
+		 *
+		 * Because of that, we MUST set pín_val for PULL_UP/DOWN.
+		 */
+		drv_mode |= CY_GPIO_DM_HIGHZ;
+	} else if (flags & GPIO_OUTPUT) {
+		drv_mode |= CY_GPIO_DM_STRONG_IN_OFF;
+		pin_val = (flags & GPIO_OUTPUT_INIT_HIGH) ? 1 : 0;
+	}
```

Comment:
I believe this will not work. To configure pin as an open source output we would pass (GPIO_OUTPUT | GPIO_OPEN_SOURCE) as the flag parameter. The `drv_mode` value which was configured by `if (flags & GPIO_SINGLE_ENDED) {` will be overwritten here.

The code should also work in case we pass `(GPIO_OUTPUT | GPIO_INPUT)`, which is valid.

Reply:
> I believe this will not work. To configure pin as an open source output we would pass (GPIO_OUTPUT | GPIO_OPEN_SOURCE) as the flag parameter. The `drv_mode` value which was configured by `if (flags & GPIO_SINGLE_ENDED) {` will be overwritten here.
> 
> The code should also work in case we pass `(GPIO_OUTPUT | GPIO_INPUT)`, which is valid.

Questions:
1) How differentiate a true Analog pin from Digital pin?
2) What `(GPIO_OUTPUT | GPIO_INPUT)` should represent?

Reply:
>     1. How differentiate a true Analog pin from Digital pin?

I assume this is a question on how to configure signal routing so the pin which can be connected to an analog block or GPIO is connected to the latter? This may need support from pinctrl, however pinctrl driver is still work in progress.

>     2. What `(GPIO_OUTPUT | GPIO_INPUT)` should represent?

This represents an in/out pin. That is, pin which can be simultaneously used as input and output. In a typical case when the output is configured in push-pull mode having the possibility to read back its input level is not really useful. We drive it and if at some point we need to read back its state it's almost always better to save the state internally. However, an in/out pin configuration becomes critical when it's configured in open-source, open-drain mode. In this case what we drive an what we read back may differ.

Reply:
Thank you @mnkp !

I reviewed and PSoC-6 bring some challenges like simultaneous PULL-UP/DOWN. I noted this is an invalid condition at GPIO driver but I'll keep implemented here since it is a valid use case. I saw that ST define Analog by not (In & Out) and I follow same here since pin is always connected. The PSoC-6 have 3 bits to define drv_mode and all valid combination were implemented. Because of the 3 bits, the GPIO_OUTPUT must select only strong or single ended. The pull configuration is only valid if not GPIO_OUTPUT. All output modes can have input buffer enabled or not.

All the cosmetic were fixed and tests were enabled at board yaml files.